runners/core-java/src/main/java/org/apache/beam/runners/core/triggers/TriggerStateMachines.java

```diff
@@ -0,0 +1,215 @@
+        } else {
+          throw new RuntimeException(exc.getCause());
+        }
+      } catch (IllegalAccessException exc) {
+        throw new IllegalStateException(
+            String.format("Internal error: could not invoke %s", evaluationMethod));
+      }
+    }
+
+    private Method getEvaluationMethod(Class<?> clazz) {
+      Method evaluationMethod;
+      try {
+        return getClass().getDeclaredMethod("evaluateSpecific", clazz);
+      } catch (NoSuchMethodException exc) {
+        throw new UnsupportedOperationException(
+            String.format(
+                "Cannot translate trigger class %s to a state machine.", clazz.getCanonicalName()),
+            exc);
+      }
+    }
+
+    private TriggerStateMachine evaluateSpecific(DefaultTrigger v) {
+      return DefaultTriggerStateMachine.of();
+    }
+
+    private OnceTriggerStateMachine evaluateSpecific(AfterWatermark.FromEndOfWindow v) {
+      return AfterWatermarkStateMachine.pastEndOfWindow();
+    }
+
+    private OnceTriggerStateMachine evaluateSpecific(NeverTrigger v) {
+      return NeverStateMachine.ever();
+    }
+
+    private OnceTriggerStateMachine evaluateSpecific(AfterSynchronizedProcessingTime v) {
+      return new AfterSynchronizedProcessingTimeStateMachine();
+    }
+
+    private OnceTriggerStateMachine evaluateSpecific(AfterFirst v) {
+      List<OnceTriggerStateMachine> subStateMachines =
+          Lists.newArrayListWithCapacity(v.subTriggers().size());
+      for (Trigger subtrigger : v.subTriggers()) {
+        subStateMachines.add(stateMachineForOnceTrigger((OnceTrigger) subtrigger));
+      }
+      return AfterFirstStateMachine.of(subStateMachines);
+    }
+
+    private OnceTriggerStateMachine evaluateSpecific(AfterAll v) {
+      List<OnceTriggerStateMachine> subStateMachines =
+          Lists.newArrayListWithCapacity(v.subTriggers().size());
+      for (Trigger subtrigger : v.subTriggers()) {
+        subStateMachines.add(stateMachineForOnceTrigger((OnceTrigger) subtrigger));
+      }
+      return AfterAllStateMachine.of(subStateMachines);
+    }
+
+    private OnceTriggerStateMachine evaluateSpecific(AfterPane v) {
+      return AfterPaneStateMachine.elementCountAtLeast(v.getElementCount());
+    }
+
+    private TriggerStateMachine evaluateSpecific(AfterWatermark.AfterWatermarkEarlyAndLate v) {
+      AfterWatermarkStateMachine.AfterWatermarkEarlyAndLate machine =
+          AfterWatermarkStateMachine.pastEndOfWindow()
+              .withEarlyFirings(stateMachineForOnceTrigger(v.getEarlyTrigger()));
+
+      if (v.getLateTrigger() != null) {
+        machine = machine.withLateFirings(stateMachineForOnceTrigger(v.getLateTrigger()));
+      }
+      return machine;
+    }
+
+    private TriggerStateMachine evaluateSpecific(AfterEach v) {
+      List<TriggerStateMachine> subStateMachines =
+          Lists.newArrayListWithCapacity(v.subTriggers().size());
+
+      for (Trigger subtrigger : v.subTriggers()) {
```

Comment:
Yeah, as mentioned elsewhere, you never need to access `subTriggers` from the actual `Trigger` class.